src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberHostContext
+ * @flow
+ */
+
+'use strict';
+
+import type { Fiber } from 'ReactFiber';
+import type { HostConfig } from 'ReactFiberReconciler';
+
+export type HostContext<C, CX> = {
+  getRootHostContainer() : C,
+  getHostContext() : CX | null,
+
+  pushHostContext(fiber : Fiber) : void,
+  popHostContext(fiber : Fiber) : void,
+
+  pushHostContainer(container : C) : void,
+  popHostContainer() : void,
+  resetHostContainer() : void,
+};
+
+module.exports = function<T, P, I, TI, C, CX>(
+  config : HostConfig<T, P, I, TI, C, CX>
+) : HostContext<C, CX> {
+  const {
+    getChildHostContext,
+  } = config;
+
+  type ContainerState = {
+    rootInstance : C,
+    contextFibers : Array<Fiber | null> | null,
+    contextValues : Array<CX | null> | null,
+    contextDepth: number
+  };
+
+  function createContainerState(rootInstance) {
+    return {
+      rootInstance,
+      contextFibers: null,
+      contextValues: null,
+      contextDepth: -1,
+    };
+  }
+
+  // If we meet any portals, we'll go deeper.
+  let containerStack : Array<ContainerState> = [];
+  let containerDepth : number = -1;
+
+  function getRootHostContainer() : C {
+    if (containerDepth === -1) {
+      throw new Error('Expected to find a root container.');
+    }
+    const containerState = containerStack[containerDepth];
+    const {rootInstance} = containerState;
+    return rootInstance;
+  }
+
+  function pushHostContainer(portalHostContainer) {
+    containerDepth++;
+    if (containerDepth === containerStack.length) {
+      containerStack[containerDepth] = createContainerState(portalHostContainer);
+    }
+  }
+
+  function popHostContainer() {
+    if (containerDepth === -1) {
+      throw new Error('Already reached the root.');
+    }
+    containerDepth--;
+  }
+
+  function getHostContext() : CX | null {
+    if (containerDepth == null) {
+      throw new Error('Expected to find a root container.');
+    }
+    const containerState = containerStack[containerDepth];
+    const {contextDepth, contextValues} = containerState;
+    if (contextDepth === -1) {
+      return null;
+    }
+    if (contextValues == null) {
+      throw new Error('Expected context values to exist.');
+    }
+    return contextValues[contextDepth];
+  }
+
+  function pushHostContext(fiber : Fiber) : void {
+    const parentHostContext = getHostContext();
+    const currentHostContext = getChildHostContext(parentHostContext, fiber.type);
+    if (parentHostContext === currentHostContext) {
```

Comment:
Here's the goal: This whole thing needs to be inlinable so that this check doesn't cause extra cost for non-context changing components. Currently this function is called multiple times which makes this a tradeoff a compiler might not do. Also getHostContext is similar. And regardless this needs at least two checks, one to compare the type string and one to compare the object pointers.